### PR TITLE
Return isolated windows from Window* methods

### LINKED ISSUE #652: Windows of Window(Left|Right) are not isolated from each other
List returned by Window* methods should be isolated from each other.

With current implementation, modifying one window will modify the next windows accordingly:

```C#
var sequence = Enumerable.Repeat(0, 3);
foreach (var window in sequence.Window(2))
{
  window[1] = 1; // modify the last element

  // on second iteration window[0] is 1 and this fail
  Assert.That(window[0], Is.EqualTo(0));
}
```

### PR DESCRIPTION
**Pull request by Orace**
This allow tests from #653 to pass and fix #652

### CODE REVIEW

**Comment by Orace** [comment]
@atifaziz can you review this please.
Fix #652

File: MoreLinq.Test/WindowLeftTest.cs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+using NUnit.Framework.Interfaces;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowLeftTest.cs]: Move under namespace declaration, together with other imports in the file.

File: MoreLinq.Test/WindowLeftTest.cs [hunk 2]
```diff
@@ -76,5 +123,25 @@ public void WindowLeftWithWindowSizeSmallerThanSequence()
...
+            {
+                new {Source = Enumerable.Range(0, 4), Size = 1, Result = new[] {Seq(0), Seq(1), Seq(2), Seq(3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 2, Result = new[] {Seq(0, 1), Seq(1, 2), Seq(2, 3), Seq(3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 3, Result = new[] {Seq(0, 1, 2), Seq(1, 2, 3), Seq(2, 3), Seq(3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 4, Result = new[] {Seq(0, 1, 2, 3), Seq(1, 2, 3), Seq(2, 3), Seq(3)}}
+            }
+            select new TestCaseData(e.Source, e.Size).Returns(e.Result);
+
+        [Test, TestCaseSource(nameof(TestData))]
+        public IEnumerable<IEnumerable<int>> TestWindowLeftOnKnownResults(IEnumerable<int> sequence, int sizes)
```
Comment (atifaziz) [inline MoreLinq.Test/WindowLeftTest.cs]: What has this test got to do with isolation of windows? Does it belong in another PR?

File: MoreLinq.Test/WindowRightTest.cs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+using NUnit.Framework.Interfaces;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowRightTest.cs]: Move under namespace declaration, together with other imports in the file.

File: MoreLinq.Test/WindowRightTest.cs [hunk 4]
```diff
@@ -76,5 +123,25 @@ public void WindowRightWithWindowSizeSmallerThanSequence()
...
+            {
+                new {Source = Enumerable.Range(0, 4), Size = 1, Result = new[] {Seq(0), Seq(1), Seq(2), Seq(3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 2, Result = new[] {Seq(0), Seq(0, 1), Seq(1, 2), Seq(2, 3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 3, Result = new[] {Seq(0), Seq(0, 1), Seq(0, 1, 2), Seq(1, 2, 3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 4, Result = new[] {Seq(0), Seq(0, 1), Seq(0, 1, 2), Seq(0, 1, 2, 3)}}
+            }
+            select new TestCaseData(e.Source, e.Size).Returns(e.Result);
+
+        [Test, TestCaseSource(nameof(TestData))]
+        public IEnumerable<IEnumerable<int>> TestWindowRightOnKnownResults(IEnumerable<int> sequence, int sizes)
```
Comment (atifaziz) [inline MoreLinq.Test/WindowRightTest.cs]: What has this test got to do with isolation of windows? Does it belong in another PR?

File: MoreLinq.Test/WindowTest.cs [hunk 5]
```diff
@@ -114,5 +162,24 @@ public void TestWindowWindowsImmutability()
...
+            {
+                new {Source = Enumerable.Range(0, 4), Size = 1, Result = new[] {Seq(0), Seq(1), Seq(2), Seq(3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 2, Result = new[] {Seq(0, 1), Seq(1, 2), Seq(2, 3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 3, Result = new[] {Seq(0, 1, 2), Seq(1, 2, 3)}},
+                new {Source = Enumerable.Range(0, 4), Size = 4, Result = new[] {Seq(0, 1, 2, 3)}}
+            }
+            select new TestCaseData(e.Source, e.Size).Returns(e.Result);
+
+        [Test, TestCaseSource(nameof(TestData))]
+        public IEnumerable<IEnumerable<int>> TestWindowOnKnownResults(IEnumerable<int> sequence, int sizes)
```
Comment (atifaziz) [inline MoreLinq.Test/WindowTest.cs]: What has this test got to do with isolation of windows? Does it belong in another PR?

File: MoreLinq.Test/WindowTest.cs [hunk 6]
```diff
@@ -1,3 +1,6 @@
+using System.Collections.Generic;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowTest.cs]: Move these under namespace declaration, together with other imports in the file.

**Comment by Orace** [comment]
> The only trouble with the new correction is that it now eagerly allocates and prepares the next window

this should be fixed in #656

**Comment by atifaziz** [comment]
> And when the sequence reach its end, all allocated array have been passed to the caller

How do you know you'll reach the end of the sequence. The caller is in control of asking the next window and may stop iteration earlier due to some condition.

File: MoreLinq.Test/WindowLeftTest.cs [hunk 7]
```diff
@@ -19,6 +19,7 @@ namespace MoreLinq.Test
 {
     using System.Collections.Generic;
     using NUnit.Framework;
+    using NUnit.Framework.Interfaces;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowLeftTest.cs]: Unused:

```suggestion
```

File: MoreLinq.Test/WindowRightTest.cs [hunk 8]
```diff
@@ -19,6 +19,7 @@ namespace MoreLinq.Test
 {
     using System.Collections.Generic;
     using NUnit.Framework;
+    using NUnit.Framework.Interfaces;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowRightTest.cs]: Unused:

```suggestion
```

File: MoreLinq.Test/WindowTest.cs [hunk 9]
```diff
@@ -17,7 +17,9 @@
 
 namespace MoreLinq.Test
 {
+    using System.Collections.Generic;
     using NUnit.Framework;
+    using NUnit.Framework.Interfaces;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowTest.cs]: Unused:

```suggestion
```

File: MoreLinq.Test/WindowTest.cs [hunk 10]
```diff
@@ -17,7 +17,9 @@
 
 namespace MoreLinq.Test
 {
+    using System.Collections.Generic;
```
Comment (atifaziz) [inline MoreLinq.Test/WindowTest.cs]: Unused:

```suggestion
```

File: MoreLinq/Window.cs [hunk 11]
```diff
@@ -52,21 +52,20 @@ public static IEnumerable<IList<TSource>> Window<TSource>(this IEnumerable<TSour
                 if (i < size)
                     yield break;
 
-                // return the first window (whatever size it may be)
-                yield return window;
-
-                // generate the next window by shifting forward by one item
-                while (iter.MoveNext())
+                while(iter.MoveNext())
```
Comment (atifaziz) [inline MoreLinq/Window.cs]: Space:

```suggestion
                while (iter.MoveNext())
```

**Comment by Orace** [comment]
> > And when the sequence reach its end, all allocated array have been passed to the caller
> 
> How do you know you'll reach the end of the sequence. The caller is in control of asking the next window and may stop iteration earlier due to some condition.

It's an hypothesis.
For a partially enumerated sequence both algorithm instantiate the same number of array  (of the size of the window).
For a fully enumerated sequence your algorithm instantiate one array more than the actual code.

**Comment by atifaziz** [comment]
> @atifaziz I do a re-read of the code

I re-read as well, and in detail this time, and it looks good! 💯
